### PR TITLE
sway: respect systemdSupport and dbusSupport

### PR DESCRIPTION
**Pull request by ghost**
This patch causes sway to honor the `systemdSupport` and `dbusSupport` attrs.

The default behavior (if these attrs are undefined) is unchanged.

I have verified a full rebuild (from the bootstrap tarball) of this PR plus https://github.com/NixOS/nixpkgs/pull/160967 plus https://github.com/NixOS/nixpkgs/pull/161017 plus systemdSupport=false on top of https://github.com/NixOS/nixpkgs/commit/65c8149cb14200c172782ede4219f53ee1d751dd.

~~NOTE: I was unable to test this patch against `master` because it has upgraded to sway-1.7, which requires meson-60.0.  As the nixpkgs meson provider explains it, "Meson is a delicate package.".  I wasn't able to backport meson-60 to nixpkgs-21.11 to allow for testing.~~

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by alyssais** [comment]
This doesn't evaluate because you haven't added the basu dependency.  (Do we even have a package?)

But either way, I think systemd will still be linked by dependencies?  Do they all have options to disable systemd already?  Otherwise I'm not sure what the benefit is.

**Comment by a-m-joseph** [comment]
> This doesn't evaluate because you haven't added the basu dependency.

Really?  It evaluates for me against `master` and builds against 21.11.  I'm trying a full build now against `master` but that's like 50 derivations, will take a few hours.

What error message did you get?

> (Do we even have a package?)

No, that's just sway's goofy way of saying "I don't want a session-bus".  Check meson_options.txt, there isn't a "none" option.  The way you write "none" is "basu".  I don't think it makes any sense either, but I found it on some mailing list somewhere and it definitely works; I've been running this on all my machines (laptop, two workstations) for two months now.
 
> But either way, I think systemd will still be linked by dependencies?

It definitely isn't linked by dependencies on my system.  But I am not done upstreaming patches.  Please be patient.

The fact that I was able to pull this off with so little effort was what sold me on nix.

**Comment by a-m-joseph** [comment]
I've pushed an additional patch onto this branch that should make it easier to understand sway's awkward configuration option naming scheme.  Hopefully it is more understandable.

**Comment by a-m-joseph** [comment]
> > But either way, I think systemd will still be linked by dependencies?
> 
> It definitely isn't linked by dependencies on my system. But I am not done upstreaming patches. Please be patient.
> 
> The fact that I was able to pull this off with so little effort was what sold me on nix.

Okay, I've submitted 12 pull requests, all very short.

If you apply all 12, and add the four lines below to your `overlays.nix`, you'll get a fully-functional linux userspace (sway, ungoogled-chromium, firefox, qutebrowser, emacs, alacritty, etc) without systemd anywhere, not even a compile-time dependency.  Verified with `nix-tree` on my "myPackages" derivation.

```
  (final: prev: { systemdMinimal = prev.eudev; })
  (final: prev: { systemd = prev.systemdMinimal; })
  (final: prev: { udev = prev.systemdMinimal; })
  (final: prev: { polkit = null; pcsclite = null; })
```

**It *is* possible.**

The same is possible for dbus.  Let's work through the `systemdSupport` PRs first.

**Comment by a-m-joseph** [comment]
I'm almost finished with a full rebuild (from the bootstrap tarball) of this PR plus  #160967 plus `systemdSupport=false` on top of 65c8149cb14200c172782ede4219f53ee1d751dd.

File: pkgs/applications/window-managers/sway/default.nix
Comment (primeos) [inline pkgs/applications/window-managers/sway/default.nix:59]: This makes sense to keep this a refactoring but for better readability I'd prefer to add this at the end instead of splitting up the list.

File: pkgs/applications/window-managers/sway/default.nix
Comment (primeos) [inline pkgs/applications/window-managers/sway/default.nix:18]: Note: Not sure if that will be future proof but that doesn't mean that we cannot use that trick for now. I didn't change the source-code though (at least not yet). It would probably make sense to contact upstream regarding a `-Dsd-bus-provider=none` option or something similar (they might not like it though if they see it as a hard requirement - one should check first if this was already discussed).

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (primeos) [inline pkgs/applications/window-managers/sway/wrapper.nix:27]:
```suggestion
   '' + lib.optionalString dbusSupport ''
```

Nit: IMO it's more readable that way (and AFAIK that's the usual style).

**Review by primeos — CHANGES_REQUESTED** [review]
Hi, thanks for your PR. I'll add some quick feedback from my side. Overall it looks good to me (and thanks for the comments regarding that trick!), just a few last nits.

First commit: I'd use `sway: Add systemdSupport and dbusSupport` or `sway: Make systemd and D-Bus optional` ("honor" seems inappropriate to me as those parameters didn't exist before). A few sentences in the body for the motivation/use-case would also be very nice and useful. I assume you're using (Free)BSD or another systemd-less OS?
Second commit: I assume it'll be squashed (otherwise it needs to be reworded).

> I also removed the -Ddefault-wallpaper=false; that is part of a separate (not yet submitted) patch to break the sway->swaybg dependency. Sorry about that.

You should submit that part with the other patch, if possible.

> Let me know if I should rebase this branch into a single commit.

Yes, that seems best to me.

File: pkgs/applications/window-managers/sway/default.nix
Comment (a-m-joseph) [reply pkgs/applications/window-managers/sway/default.nix:18]: Done: https://github.com/swaywm/sway/issues/6843

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (a-m-joseph) [reply pkgs/applications/window-managers/sway/wrapper.nix:27]: Done, thanks (will push momentarily).  Yes, I agree, it's more readable that way.

File: pkgs/applications/window-managers/sway/default.nix
Comment (a-m-joseph) [reply pkgs/applications/window-managers/sway/default.nix:59]: My instinct is generally to maintain the ordering of the buildInputs in the default case, to avoid changing the derivation-hash and therefore triggering unnecessary rebuilds for people who do not customize the value of `dbusSupport`.

I guess in this case it doesn't matter much since very few packages depend on `sway`.  I've implemented your recommendation and will push a new commit in a moment.

**Comment by a-m-joseph** [comment]
Thanks for the review!  I have applied your three recommended changes and pushed a collapsed single-commit PR.
 
> First commit: I'd use `sway: Add systemdSupport and dbusSupport` or `sway: Make systemd and D-Bus optional` ("honor" seems inappropriate to me as those parameters didn't exist before). 

Actually they did!  They are top-level parameters that you can set in `config.nix` -- this is where I set them.  That's why it's important to use the same name `dbusSupport` across packages, rather than some packages calling it `enableDbus` or `useDbus` or something else.

I've changed the title to "respect" in case that carries less connotation?  Another possibility would be "don't ignore".  But it is a globally-configurable option that long predated this PR.

> A few sentences in the body for the motivation/use-case would also be very nice and useful.

I'm have very deliberately not done that, because for some reason people get very emotional about systemd.

> I assume you're using (Free)BSD or another systemd-less OS?

I don't think my personal computing setup belongs in the commit message, but since you asked:

I've been running systemd-less, dbus-less Linux since the 1990s, and have used `runit` for process supervision since before systemd existed (I'm currently migrating from `runit` to `s6`, which is quite similar to `runit` but has many improvements).  It is sort of incredible how many people think that this is somehow impossible or defies the laws of physics.  When people express disbelief, I have to carefully distinguish between those people who are genuinely curious about how this works and other people who get very angry and upset at me for daring to use Linux without systemd and dbus.  I have a feeling some of the latter kind will probably come out of the woodwork to heckle me now that I've confessed this.

Currently I use this setup on my laptop, two workstations, eight or nine headless routers, and two headless build-farm machines.  I like it a lot.

File: pkgs/applications/window-managers/sway/default.nix
Comment (Synthetica9) [inline pkgs/applications/window-managers/sway/default.nix:19]: Shouldn't this be the other way around? 

`trayEnabled -> systemdSupport && dbusSupport`

Currently as it written it would imply that _either_ systemd _or_ dbus is required for tray support, isn't it both?

Comment (a-m-joseph) [reply pkgs/applications/window-managers/sway/default.nix]: Wow, thanks.  I got the [contrapositive](https://en.wikipedia.org/wiki/Contraposition) wrong.  Corrected in latest push.

Comment (Synthetica9) [reply pkgs/applications/window-managers/sway/default.nix]: (even if it's either-or I still think `trayEnabled -> systemdSupport || dbusSupport` would be clearer)

File: pkgs/applications/window-managers/sway/default.nix
Comment (a-m-joseph) [reply pkgs/applications/window-managers/sway/default.nix:18]: Hrm, well, the [reply](https://github.com/swaywm/sway/issues/6843#issuecomment-1047288761) was sort of "working as intended".

It looks like there are five possibilities not including `basu` which is not (currently) in nixpkgs:

1. libsystemd (and dbus) and tray=enabled
2. libsystemd (and dbus) and tray=disabled
3. tray=disabled
4. elogind (and dbus) and tray=enabled
5. elogind (and dbus) and tray=disabled

The last two possibilities are not (currently) supported by nixpkgs' sway derivation.

I have updated the PR to include a link to the sway maintainers' comment, rewritten the `-Dsd-bus-provider=` flag more compactly, and factored out a definition of `sd_bus_provider` so I could place that definition close to the `assert` that explains the whole situation up at the top of the file (otherwise somebody looking only at `mesonFlags` would think to themselves "what is this `basu` stuff?").

I think that if `elogindSupport` were to be added at some point in the future, the `assert` line would change from

```
assert trayEnabled -> systemdSupport && dbusSupport;
```

to

```
assert trayEnabled -> (systemdSupport || elogindSupport) && dbusSupport;
```

and an `else if` would be added to the `sd_bus_provider` definition.

**Comment by a-m-joseph** [comment]
Latest push is explained [here](https://github.com/NixOS/nixpkgs/pull/160972#discussion_r811542665).

**Comment by a-m-joseph** [comment]
Latest push adds a commit message only -- no other changes.

**Comment by Synthetica9** [comment]
Very nice. We should probably extend the NixOS test to test a non-dbus
non-systemd version as well, but I don't know if you have experience with
that?

**Comment by a-m-joseph** [comment]
Unfortunately I have not yet installed NixOS, so it is definitely beyond my abilities at the moment.

Learning NixOS is on my to-do list, but there are a few major items ahead of it -- the big ones being finishing the mips64 bootstrap and setting up a hydra instance on mips64.  Those will take some time, however, so I would prefer that it not block this.

**Comment by a-m-joseph** [comment]
> NixOS test to test a non-dbus non-systemd version as well

I was under the impression that NixOS was rather inextricably intertwined with systemd.  I would happy to be wrong about that.

I can certainly see why NixOS needs process supervision.

If there are any efforts to make NixOS work with a different process supervisor like (s6, runit, daemontools) I would be interested in contributing to those efforts.  But I had been told that something like that was not particularly realistic, and would not be well received by the NixOS community.  I am not looking to stir up trouble.

**Comment by Synthetica9** [comment]
> I was under the impression that NixOS was rather inextricably intertwined with systemd. I would happy to be wrong about that.

Well yes, but you can still test that the non-systemd version of sway works even if you are running on systemd.

**Comment by a-m-joseph** [comment]
Ping...

(I implemented the three changes requested by @primeos, but github still says "1 change requested: primeos requested changes" -- I guess this is because I wrote and committed the changes outside the github UI rather than using the web interface?)

File: pkgs/applications/window-managers/sway/default.nix
Comment (Artturin) [inline pkgs/applications/window-managers/sway/default.nix:12]:
```suggestion
, dbusSupport ? true
, dbus
```
and then it should be removed above

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (Artturin) [inline pkgs/applications/window-managers/sway/wrapper.nix:27]: won't this mean that if dbus support is not enabled then the baseWrapper won't have a `exec ${sway}/bin/sway "$@"`

**Review by Artturin — CHANGES_REQUESTED** [review]
`sway-unwrapped` builds but `sway` won't `sway> /nix/store/rj7vaqa0a2ra6dyz4rbkmj8d126dfwaj-swayif: No such file or directory`

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (a-m-joseph) [reply pkgs/applications/window-managers/sway/wrapper.nix:27]: Yikes, thanks for catching this.  I don't use the wrapper script which is why I didn't notice.  Sorry about that; will push another commit to fix this.

File: pkgs/applications/window-managers/sway/default.nix
Comment (Artturin) [reply pkgs/applications/window-managers/sway/default.nix:12]: same for your dbusSupport PRs btw

Comment (a-m-joseph) [reply pkgs/applications/window-managers/sway/default.nix:12]: Does systemd work on anything other than Linux?  I didn't think so but I could be wrong.

There are seven other packages with `systemdSupport` defaulting to `stdenv.isLinux`, and only one which defaults to `true` is some kind of tray widget for KDE.

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (a-m-joseph) [reply pkgs/applications/window-managers/sway/wrapper.nix]: Okay, pushed.

I tried to write the conditional logic in the most maintainable way I could.  The doubly-nested antiquotation isn't pretty, but wrapping the entire `if`..`else`..`fi` block in `lib.optional` and then adding a completely separate code path seemed even worse.

File: pkgs/applications/window-managers/sway/default.nix
Comment (a-m-joseph) [reply pkgs/applications/window-managers/sway/default.nix:12]: Also I *think* that this way Darwin users should be able to just `nix-env -iA sway` and run it using `WLR_BACKENDS=headless` (and RDP) or `WLR_BACKENDS=x11` (and XQuartz).  But I don't have a Mac so I can't try this.

Comment (Artturin) [reply pkgs/applications/window-managers/sway/default.nix:12]: i think the github syntax highlighter does that because the comma color in the nix syntax highlighting is similar to the removed color

**Review by primeos — APPROVED** [review]
> Ping...

Sorry about the delay and thanks for re-pinging me. I do sometimes unfortunately loose track of PRs due to lack of time.

Everything LGTM now and I like that `${if !dbusSupport then "" else "${dbus}/bin/dbus-run-session"}` approach. And a very beautiful commit message! :) (I just hope it didn't sound like I wanted that much text but it's definitely nice.)

> They are top-level parameters that you can set in ~/.config/nixpkgs/config.nix -- this is where I set them. That's why it's important to use the same name dbusSupport across packages, rather than some packages calling it enableDbus or useDbus or something else.

My bad, thanks for clarifying. I didn't knew that this works now. IIRC this trick only used to work with `config.*` attributes (e.g., `config.pulseaudio`) so it looks like the PRs to improve that made it through (although our overriding system still isn't ideal - there was e.g. the idea to extend the Nix language to better support configurations: https://gist.github.com/edolstra/29ce9d8ea399b703a7023073b0dbc00d).

Anyway, the commit message makes sense then :)

> I don't think my personal computing setup belongs in the commit message, but since you asked:

Sorry, I mainly wanted one brief example in the commit message as not everyone might get why we need this change. Thanks for sharing your cool setup though! I found that very interesting :) And thanks for upstreaming your patches.

I'll give it a quick test and will squash and merge.
(Edit: Not quite sure if we should also propagate `dbusSupport` through via `sway-unwrapped.override { inherit isNixOS enableXWayland; };` but we could always re-visit that later if necessary/useful and it should be good to go for now. Another nit would be `trayEnabled ? systemdSupport && dbusSupport`.)

**Comment by primeos** [comment]
Merged, thanks for your PR :)

**Comment by a-m-joseph** [comment]
Thank you!

File: pkgs/applications/window-managers/sway/wrapper.nix
Comment (Artturin) [inline pkgs/applications/window-managers/sway/wrapper.nix:31]: optionalString would have been better https://github.com/NixOS/nixpkgs/blob/f8073ff44c8d06a354622cf5ced7d9d854c61f2a/lib/strings.nix#L198-L213

Comment (a-m-joseph) [reply pkgs/applications/window-managers/sway/wrapper.nix:31]: Thank you; I will use that next time.

**Comment by a-m-joseph** [comment]
> (Edit: Not quite sure if we should also propagate `dbusSupport` through via `sway-unwrapped.override { inherit isNixOS enableXWayland; };` but we could always re-visit that later if necessary/useful and it should be good to go for now.

I'm not yet using NixOS (just nixpkgs) so I at the moment I can't test that properly.

> Another nit would be `trayEnabled ? systemdSupport && dbusSupport`.)

Implemented in PR #163402

**Pull request merged**
